### DISCUSSION TITLE
Supabase RLS Assistant feedback, round 2

### DISCUSSION THREAD

**Discussion by saltcod** [supabase/supabase#21882] [General]
We've recently shipped a new iteration of our RLS Assistant. 

You sent us a ton of feedback, and we read it all. We dug in to try and fix some of the issues with the iteration we launched in December. 

**What we heard:** 

- many of you liked the pre-defined templates in the old UI 
- the AI responses were hit and miss (turns out, LLMs are great with general SQL, but pretty weak with RLS policies)
- opting-in to sending their db schemas along with the prompts isn't obvious

This v2 brings templates back as the central feature, moving the AI Assistant to the background. It's still there, and it's tuned to be a little better for RLS policies this time around. But the feature now revolves around templates again. What's old is new. 

**What we'd like to know:** 

- We think this is a huge improvement over the last iteration. Do you agree? 
- Does this help you write RLS policies? Does it fit your workflow? 
- Anything you think is missing? Any templates missing? 
- How are the AI responses this time?

**What we're thinking about next:** 

We're going to let this version simmer for a bit and gather new feedback before we make any more changes, but these are under consideration: 

- allowing you to add your own templates 
- adding filtering and favouriting to the list
- letting you see all of your table's current RLS policies in one place    

![screenshot-2024-03-08-at-09 54 16](https://github.com/supabase/supabase/assets/105593/1f2f442e-790d-482c-a517-6f5b9a9aa1af)

**Notes:** 

You can enable the new UI via Feature Previews → Supabase Assistant for RLS policies

![screenshot-2024-03-08-at-09 57 00](https://github.com/supabase/supabase/assets/105593/2ed278a5-7d51-4676-946b-c522776d5a32)

This is a follow-up to the previous discussion https://github.com/orgs/supabase/discussions/19594

**Comment by eqoram** [comment]
I am still missing some kind of functionality to easily define that a policy should run in "system context" instead of "user context". This can be achieved by creating a postgres function and then calling this in the policy but it would be great to have this option directly in the editor.

**Reply by GaryAustin1** [reply]
The editor would still need to create such a security definer function as that is the only way Postgres policies can be done to bypass.

**Comment by anngbaum** [comment]
We're adding a bunch of new RLS policies right now, and our biggest concern is performance, so we're very focused on making sure we use the best practices [here](https://github.com/orgs/supabase/discussions/14576) (thank you @GaryAustin1 !) 

I took a quick glance at the new templates side, and thought that the update template seemed inconsistent with 2a as well as 5. 
<img width="908" alt="Screenshot 2024-03-08 at 12 12 49 PM" src="https://github.com/supabase/supabase/assets/29735378/399fa995-ea64-4c36-922d-24a422a87f67">

Based on my (current!) understanding of the suggestions outlined, the template should be written as:

```
on teams for update using (
  id in (
    select team_id from members where user_id = auth.uid()
  )
);
```

I could be misunderstanding, but that would be by far my biggest concern!

**Reply by GaryAustin1** [reply]
Just to add that troubleshooting guide was also rolled into the docs by Supabase and cleaned up a bit https://supabase.com/docs/guides/database/postgres/row-level-security#minimize-joins .
The thing to watch for on the method is if the in list gets to be 1000 or more then you really have to study performance closely.  But for majority of cases it is much faster.

**Reply by olirice** [reply]
When in doubt, its always a good idea to check what the explain plan says for your target user!

```sql
/*
  SETUP
*/

create table teams (
  id int primary key,
  val text
);

alter table teams enable row level security;

insert into teams(id)
select * from generate_series(1, 1000);

create table members (
  id serial primary key,
  team_id int not null references teams (id),
  user_id uuid not null
);

create index ix_members_team_id on members (team_id);
create index ix_members_user_id on members (user_id);

insert into members (team_id, user_id)
select
  t.id,
  case when mod(t.id, 30) = 0 then (select id from auth.users limit 1) else gen_random_uuid() end 
from
  teams t;


/*
  Option 1
*/

create policy "pol1"
  on teams
  for update using (
  id in (
    select team_id from members where user_id = auth.uid()
  )
);


-- Now use User Impersonation (https://supabase.com/blog/studio-introducing-assistant#user-impersonation)
-- to look at the explain plan for the user/role that will be updating the "teams" table

-- Change the query as appropriate
explain analyze update teams set val = 'foo' where id = 1;

/*
"Update on teams (cost=10.03..12.25 rows=0 width=0) (actual time=0.033..0.033 rows=0 loops=1) "
" -> Index Scan using teams_pkey on teams (cost=10.03..12.25 rows=1 width=38)
(actual time=0.032..0.032
rows=0 loops=1)"
" Index Cond: (id = 1)"
" Filter: (false AND (hashed SubPlan 2)) "
" Rows Removed by Filter: 1"
" SubPlan 2"
" -> Index Scan using ix_members_user_id on members members_1 (cost=0.30..9.67 rows=33 width=4) (never executed)"
" Index Cond: (user_id = (COALESCE(NULLIF (current_setting('request. jwt.claim.sub'::text, true), ''::text), ((NULLIF(current_setting('request jwt,claims'::text, true), "'::text))::jsonb ->› 'sub'::text)))::uuid)"
" SubPlan 1"
" -> Index Scan using ix _members_user_id on members (cost=0.30..9.67 rows=33 width=4) (never executed)"
" Index Cond: (user_id = (COALESCE(NULLIF (current_setting ('request.jwt.claim.sub'::text, true), "'::text), ((NULLIF(current_setting('request jwt.claims'::text, true), "'::text)) ::jsonb
→>> 'sub'::text))) ::uuid)"
"Planning Time: 0.772 ms"
"Execution Time: 0.173
*/
```

Once you've got the explain plan for a representative query, you can lookout for unexpectedly slow execution times or any nodes that say "sequential scan" as that will be very slow.

The policy you have above looks great as long as your `team_id` column is indexed in the `members` table.

**Reply by anngbaum** [reply]
agreed! we've definitely found `explain analyze` to be really helpful for these types of issues. I guess to be more specific – my feedback was that the template policy suggested by the new UI doesn't seem to align with best practices.  

for the `Policy with table joins` the UI suggests:
```
on teams for update using (
  auth.uid() in (
    select user_id from members where team_id = id
  )
);
```
running the (super helpful!) testing you outlined above, the cost is roughly double:
```
| QUERY PLAN                                                                                                                |
| ------------------------------------------------------------------------------------------------------------------------- |
| Update on teams  (cost=19.79..22.03 rows=0 width=0) (actual time=0.028..0.028 rows=0 loops=1)                             |
|   ->  Index Scan using teams_pkey on teams  (cost=19.79..22.03 rows=1 width=38) (actual time=0.027..0.027 rows=0 loops=1) |
|         Index Cond: (id = 1)                                                                                              |
|         Filter: (false AND (hashed SubPlan 2))                                                                            |
|         Rows Removed by Filter: 1                                                                                         |
|         SubPlan 2                                                                                                         |
|           ->  Seq Scan on members members_1  (cost=0.00..19.50 rows=5 width=16) (never executed)                          |
|                 Filter: (team_id = id)                                                                                    |
|   SubPlan 1                                                                                                               |
|     ->  Seq Scan on members  (cost=0.00..19.50 rows=5 width=16) (never executed)                                          |
|           Filter: (team_id = id)                                                                                          |
| Planning Time: 0.980 ms                                                                                                   |
| Execution Time: 0.166 ms                                                                                                  |
```

rewriting to optimize for the best practices reduces the cost (just as you tested and called out in your comment) I get:
```
| Update on teams  (cost=8.44..10.66 rows=0 width=0) (actual time=0.037..0.038 rows=0 loops=1)                                                                                                                                     
```

that change becomes magnified with more realistic data (just try adding 1000 rows to one specific team, and found the cost went up to 38 for the unoptimized query, and only 2 for the optimized query.)

I realize this is a nitpick as both policies work technically, but definitely the most useful/important feature for us is making sure that the policies are as efficient as possible!

**Reply by anngbaum** [reply]
ok one other (more straightforward) template issue that I ran into last night – the template `Enable insert for users based on user_id` suggests:
```
create policy "Enable insert for users based on user_id"
on schema_name.table_name
for insert using (
  auth.uid() = user_id
) with check (
  auth.uid() = user_id
);
```

my understanding is that `INSERT` RLS policies shouldn't have `using` – that check only applies to existing rows, i.e. `UPDATE` or `DELETE`. `INSERT` policies only need the `with check`.

relatedly, in a UI bug – the error message when I tried to run that policy for a given table didn't show up:
![Screenshot 2024-03-12 at 8 33 13 AM](https://github.com/supabase/supabase/assets/29735378/07baab30-c658-4d40-b0b3-2473550ae89d)

I only noticed the issue once I retried from the SQL editor:
![Screenshot 2024-03-12 at 8 33 33 AM](https://github.com/supabase/supabase/assets/29735378/8169d6ea-90a1-4596-9e6b-daab49246a58)

**Reply by GaryAustin1** [reply]
Commenting on the indexing of the "joined" table (members).  That is absolutely required in the un-optimized case as the query will be run once for each row in teams.  With the optimized case that is not so critical, or at least should be studied in relation to how often the joined table is updated/inserted as those will slow down with indexes.  The optimized case is set up to run the expensive query ONCE on the join table and then the query will compare a constant with the result of that query against every row in the teams table.  My guess is for the members table shown here it does not hurt to index as it is unlikely updated frequently.

**Comment by tophat1986** [comment]
Whilst lots of positives, one negative from me is that when "editing" and seeing what the current policies are, it's too hidden and makes it difficult to see the existing ones. 

It's living under this tiny tab here: 
![image](https://github.com/supabase/supabase/assets/29504604/f949dc58-76bb-4a7c-a158-627e27c6facc)

I want to open it up and see what I've currently got and then simply make changes to that.  I've had to toggle the preview off as it was better laid out before.

**Comment by LaymanTeam** [comment]
Honestly, just add more templates, taking a regex/building-block approach to policy building would be the most helpful. Alternatively, you can pass more information about the db to the LLM for creating RLS policies, because I think the issue is that you're leaving the LLM to infer a lot from what seems to be so little data. If the issue is security your team may have to simply invest more in a blackbox model to achieve this goal.

**Comment by GaryAustin1** [comment]
The ability to copy an existing tables RLS and/or create a template from it.

**Comment by andrewransom** [comment]
When the feature is generally released, please allow us the option to access the existing (old) wizard-like policy editor. It's far more intuitive for new users like me! It seems like it's either/or now depending on whether or not you have the Preview feature enabled.
